### ISSUE TITLE
Upgrading from 3.4 to 3.5.x is broken. 

### ISSUE THREAD

**Issue by newtewt** [medic/medic#5653]
**Describe the bug**
Upgrading from 3.4 to 3.5.x is broken for offline users.  When looking in the console there are lots of errors about replication and unhandled rejections.

**To Reproduce**
1. Using a instance on 3.4
2. Log in as an offline user and replicate everything to be synced up. 
3. I went off line at this point by disabling wifi on my phone and in browser ticked the offline in the network tab. 
4. Upgrade to 3.5.x
5. Come back online. 

**Expected behavior**
The app should replicate and see the upgrade and prompt the user to update to get the latest changes. Currently it is spitting a bunch of errors in the logs. 

**Logs**
```
inbox.js:3 Replication started after 54.461 seconds since previous attempt.
inbox.js:3 Error replicating from remote server n {result: {…}}
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
a.emit @ inbox.js:3
(anonymous) @ inbox.js:3
P @ inbox.js:3
N @ inbox.js:3
(anonymous) @ inbox.js:3
Promise.catch (async)
F @ inbox.js:3
pr @ inbox.js:3
vr @ inbox.js:3
from @ inbox.js:3
(anonymous) @ inbox.js:3
v @ inbox.js:3
$ @ inbox.js:3
sync @ inbox.js:3
s.replicate @ inbox.js:3
fn @ VM1936:4
i @ inbox.js:3
$eval @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
inbox.js:3 n {result: {…}} "Possibly unhandled rejection: {"result":{"ok":false,"start_time":"2019-05-09T21:21:40.564Z","docs_read":0,"docs_written":0,"doc_write_failures":0,"errors":[],"status":"aborting","end_time":"2019-05-09T21:21:40.599Z","last_seq":0}}"
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
c @ inbox.js:3
$digest @ inbox.js:3
(anonymous) @ inbox.js:3
completeTask @ inbox.js:3
(anonymous) @ inbox.js:3
setTimeout (async)
o.defer @ inbox.js:3
$evalAsync @ inbox.js:3
(anonymous) @ inbox.js:3
l @ inbox.js:3
f @ inbox.js:3
h @ inbox.js:3
reject @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
a.emit @ inbox.js:3
(anonymous) @ inbox.js:3
P @ inbox.js:3
N @ inbox.js:3
(anonymous) @ inbox.js:3
Promise.catch (async)
F @ inbox.js:3
pr @ inbox.js:3
vr @ inbox.js:3
from @ inbox.js:3
(anonymous) @ inbox.js:3
v @ inbox.js:3
$ @ inbox.js:3
sync @ inbox.js:3
s.replicate @ inbox.js:3
fn @ VM1936:4
i @ inbox.js:3
$eval @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
inbox.js:3 Translation for sync.status.undefined doesn't exist
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
Q @ inbox.js:3
nt @ inbox.js:3
W @ inbox.js:3
k @ inbox.js:3
x @ inbox.js:3
(anonymous) @ inbox.js:3
$digest @ inbox.js:3
(anonymous) @ inbox.js:3
completeTask @ inbox.js:3
(anonymous) @ inbox.js:3
setTimeout (async)
o.defer @ inbox.js:3
$evalAsync @ inbox.js:3
(anonymous) @ inbox.js:3
l @ inbox.js:3
f @ inbox.js:3
h @ inbox.js:3
reject @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
a.emit @ inbox.js:3
(anonymous) @ inbox.js:3
P @ inbox.js:3
N @ inbox.js:3
(anonymous) @ inbox.js:3
Promise.catch (async)
F @ inbox.js:3
pr @ inbox.js:3
vr @ inbox.js:3
from @ inbox.js:3
(anonymous) @ inbox.js:3
v @ inbox.js:3
$ @ inbox.js:3
sync @ inbox.js:3
s.replicate @ inbox.js:3
fn @ VM1936:4
i @ inbox.js:3
$eval @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
inbox.js:3 Error replicating to remote server n {result: {…}}
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
a.emit @ inbox.js:3
(anonymous) @ inbox.js:3
P @ inbox.js:3
N @ inbox.js:3
(anonymous) @ inbox.js:3
Promise.catch (async)
F @ inbox.js:3
pr @ inbox.js:3
vr @ inbox.js:3
to @ inbox.js:3
(anonymous) @ inbox.js:3
v @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
$digest @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
inbox.js:3 Error replicating remote server (2) [n, n]
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
$digest @ inbox.js:3
(anonymous) @ inbox.js:3
completeTask @ inbox.js:3
(anonymous) @ inbox.js:3
setTimeout (async)
o.defer @ inbox.js:3
$evalAsync @ inbox.js:3
(anonymous) @ inbox.js:3
l @ inbox.js:3
f @ inbox.js:3
h @ inbox.js:3
reject @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
a.emit @ inbox.js:3
(anonymous) @ inbox.js:3
P @ inbox.js:3
N @ inbox.js:3
(anonymous) @ inbox.js:3
Promise.catch (async)
F @ inbox.js:3
pr @ inbox.js:3
vr @ inbox.js:3
to @ inbox.js:3
(anonymous) @ inbox.js:3
v @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
$digest @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
inbox.js:3 Replication ended after 0.062 seconds with status required
inbox.js:3 n {result: {…}} "Possibly unhandled rejection: {"result":{"ok":false,"start_time":"2019-05-09T21:21:40.567Z","docs_read":0,"docs_written":0,"doc_write_failures":0,"errors":[],"status":"aborting","end_time":"2019-05-09T21:21:40.621Z","last_seq":0}}"
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
c @ inbox.js:3
$digest @ inbox.js:3
(anonymous) @ inbox.js:3
completeTask @ inbox.js:3
(anonymous) @ inbox.js:3
setTimeout (async)
o.defer @ inbox.js:3
$evalAsync @ inbox.js:3
(anonymous) @ inbox.js:3
l @ inbox.js:3
f @ inbox.js:3
h @ inbox.js:3
reject @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
a.emit @ inbox.js:3
(anonymous) @ inbox.js:3
P @ inbox.js:3
N @ inbox.js:3
(anonymous) @ inbox.js:3
Promise.catch (async)
F @ inbox.js:3
pr @ inbox.js:3
vr @ inbox.js:3
to @ inbox.js:3
(anonymous) @ inbox.js:3
v @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
(anonymous) @ inbox.js:3
$digest @ inbox.js:3
$apply @ inbox.js:3
(anonymous) @ inbox.js:3
dispatch @ inbox.js:3
v.handle @ inbox.js:3
```

**Environment**
- Instance: localhost
- Browser: chrome and android app
- Client platform: linux
- App: webapp
- Version: 3.5.x upgrade

**Comment by garethbowen** [comment]
This is likely caused by either #5361 or #5646

**Comment by garethbowen** [comment]
@newtewt Kenn made changes on the PR above to fix this issue also, so I'm assigning to you. The PR should now fix all three issues!

**Comment by kennsippell** [comment]
@garethbowen Nick tested on Friday. Looks like upgrade is working but we still aren't seeing the "upgrade now" dialog appear.

**Comment by kennsippell** [comment]
To upgrade to 3.5, we have the following requirements:

1. Clear the appcache's "master" record for the route `/medic/_design/medic/_rewrire/` or else the page will just load from cache and not get the new version
1. Trigger the "update now" prompt

In previous versions of the app, we show the "Update Ready" dialog based on the result of the appcache update. We attempt to update the appcache whenever one of the following docs change: `_design/medic`, `_design/medic-client`, `appcache`, or `settings`. 

The current solution in 3.5 is to make it "obsolete" by returning a 404 when it is requested. That doesn't trigger the update dialog. These are the events that cause the dialog to show:

### 1. AppCache manifest undergoes no change
The current manifest caches `/medic/_design/medic/_rewire/`. If we change nothing, the upgrade is blocked by this cache entry. So we have to do something with the manifest. 

### 2. AppCache manifest changes, update succeeds
"An application cache always includes at least one resource..." [[source](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache#Resources_in_an_application_cache)] due to its master entry (it caches the page that requests the manifest). So if we serve up an empty manifest, it still caches `/medic/_design/medic/_rewire/` and the update is blocked.  

I tried using the `NETWORK` section of the manifest so it would ignore the master entry, but that doesn't appear to work. So not sure how to make this event trigger + bust the cache.

### 3. AppCache update() throws an exception
We will get an update if we get an exception. For appcache: exceptions are different from errors. Errors occur if the manifest is invalid or the manifest includes a resource which isn't available.

So the change to `_design/medic-client` doc, causes the appCache to become obsolete. Then, any second change to the docs listed above (_design/medic-client or settings) triggers another appCache `update()` which causes the exception "cannot update appcache since none exists". 


## Options
Maybe there is something tricky we can do with the appcache manifest to get 2 working .. but I'm running out of ideas there.  So I think option 3 is our best option - it just requires two document changes to get the dialog... we have `_design/medic-client` already  + one other ...

1. Cause a change to the `settings` doc. Potentially with `medic-conf` step during post-deployment?
1. Cause a change to the `appcache` couchdb doc. I removed the `appcache` doc from couchdb and it no longer replicates down to the device as of 3.5. But we could undo that part of the service-worker change and replace it with some kind of tombstone.
1. Other? @garethbowen?

**Comment by kennsippell** [comment]
Discussed with @garethbowen today. 

Plan is to : 

1. Go with Option 3 above. Write a migration to change the `settings` doc in some benign way + change the `_design/medic-client` doc as part of the release (https://github.com/medic/medic/pull/5640)
1. Investigate removing the `/empty.manifest` file since it may not be required.

**Comment by kennsippell** [comment]
Option 3 doesn't work when the changes are synced at the same time. @garethbowen - new plan required!

**Comment by garethbowen** [comment]
@newtewt We've had another go at this one - if you have time it'd be great to get this ATed! Don't merge yet as there's a couple of things to clean up but nothing that should affect the functionality. Branch: `5646-update-menu`

**Comment by garethbowen** [comment]
@newtewt Code review has passed with minimal changes since your last AT. Would you like to have one more round, or shall we just merge this as is?

**Comment by newtewt** [comment]
Yeah I'll have a quick look today and then merge if I don't see anything.

**Comment by newtewt** [comment]
Checked with the latest changes and wanted to ensure the PR comments are good before merging.

**Comment by kennsippell** [comment]
Merged!
